build/shimplify.py

```diff
@@ -0,0 +1,169 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""implementation of shimplify Ant task """
+
+import errno
+import json
+import logging
+import os
+import subprocess
+
+log = logging.getLogger('shimplify')
+log.setLevel(logging.DEBUG)
+ch = logging.StreamHandler()
+ch.setLevel(logging.DEBUG)
+formatter = logging.Formatter('%(name)s - %(levelname)s - %(message)s')
+ch.setFormatter(formatter)
+log.addHandler(ch)
+
+
+def __is_enabled_property(p):
+    assert not (p is None), "Invalid property: None"
+    prop_val = project.getProperty(p)
+    return str(True).lower() == prop_val
+
+
+def __is_enabled_attr(attr):
+    assert not (attr is None), "Invalid attribute: None"
+    attr_val = attributes.get(attr)
+    return not (attr_val is None) and __is_enabled_property(attr_val)
+
+
+
+task_enabled = __is_enabled_attr('if')
+
+# should we move files?
+move_files = __is_enabled_property('shimplifyMove')
+
+# allowed to overwrite the existing comment
+over_shimplify = __is_enabled_property("overShimplify")
+
+
+def __upsert_shim_json(filename, bv_list):
+    with open(filename, 'r') as file:
+        contents = file.readlines()
+    tag = 'spark-rapids-shim-json-lines'
+    shim_comment = []
+    shim_comment.append('/*** ' + tag)
+    for bv in bv_list:
+        shim_comment.append(json.dumps({'spark' :  bv}))
+    shim_comment.append(tag + ' ***/')
+    __delete_prior_comment_if_allowed(contents, tag, filename)
+    log.debug("Inserting comment %s to %s", shim_comment, filename)
+    package_line = next(i for i in xrange(len(contents)) if str(contents[i]).startswith('package'))
+    log.debug("filename=%s package_line_number=%d", filename, package_line)
+    shim_comment_str = '\n'.join(shim_comment) + '\n'
+    contents.insert(package_line, shim_comment_str)
+    with open(filename, 'w') as file:
+        file.writelines(contents)
+
+
+def __delete_prior_comment_if_allowed(contents, tag, filename):
+    opening_shim_comment_line = None
+    closing_shim_comment_line = None
+    try:
+        opening_shim_comment_line = next(i for i in xrange(len(contents)) if tag in str(contents[i]))
+        closing_shim_comment_line = next(i for i in xrange(opening_shim_comment_line + 1, len(contents)) if tag in str(contents[i]))
+    except StopIteration as si_exc:
+        if not (opening_shim_comment_line is None) and (closing_shim_comment_line is None):
```

Comment:
Is it valid case for `opening_shim_comment_line` to be None and `closing_shim_comment_line` to be not None?
If not then just have an xor instead?

Reply:
this line deals with the case `opening_shim_comment_line` is not a None whereas `closing_shim_comment_line` is None (missing). Can you elaborate on the suggestion?

Reply:
I think the confusion was my n00b placement of `not`